File: backend/src/main/java/com/employment/system/service/AttendanceService.java
package com.employment.system.service;

import com.employment.system.model.Attendance;
import com.employment.system.repository.AttendanceRepository;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;
import java.time.LocalDate;
import java.util.List;
import java.util.Optional;

@Service
@RequiredArgsConstructor
public class AttendanceService {

    private final AttendanceRepository attendanceRepository;

    public List<Attendance> getAllAttendance() {
        return attendanceRepository.findAll();
    }

    public Optional<Attendance> getAttendanceById(Long id) {
        return attendanceRepository.findById(id);
    }

    public List<Attendance> getAttendanceByEmployee(Long employeeId) {
        return attendanceRepository.findByEmployeeId(employeeId);
    }

    public List<Attendance> getAttendanceByDate(LocalDate date) {
        return attendanceRepository.findByDate(date);
    }

    public List<Attendance> getAttendanceByDateRange(Long employeeId, LocalDate startDate, LocalDate endDate) {
        return attendanceRepository.findByEmployeeIdAndDateBetween(employeeId, startDate, endDate);
    }

    @Transactional
    public Attendance markAttendance(Attendance attendance) {
        Optional<Attendance> existing = attendanceRepository.findByEmployeeIdAndDate(
            attendance.getEmployee().getId(), 
            attendance.getDate()
        );
        
        if (existing.isPresent()) {
            throw new RuntimeException("Attendance already marked for this date");
        }
        
        return attendanceRepository.save(attendance);
    }

    @Transactional
    public Attendance updateAttendance(Long id, Attendance attendanceDetails) {
        Attendance attendance = attendanceRepository.findById(id)
                .orElseThrow(() -> new RuntimeException("Attendance not found"));

        attendance.setCheckInTime(attendanceDetails.getCheckInTime());
        attendance.setCheckOutTime(attendanceDetails.getCheckOutTime());
        attendance.setStatus(attendanceDetails.getStatus());
        attendance.setRemarks(attendanceDetails.getRemarks());

        return attendanceRepository.save(attendance);
    }

    @Transactional
    public void deleteAttendance(Long id) {
        Attendance attendance = attendanceRepository.findById(id)
                .orElseThrow(() -> new RuntimeException("Attendance not found"));
        attendanceRepository.delete(attendance);
    }
}

File: frontend/src/services/attendanceService.js
import api from './api';

const ATTENDANCE_API = '/attendance';

export const getAllAttendance = async () => {
  const response = await api.get(ATTENDANCE_API);
  return response.data;
};

export const getAttendanceById = async (id) => {
  const response = await api.get(`${ATTENDANCE_API}/${id}`);
  return response.data;
};

export const getAttendanceByEmployee = async (employeeId) => {
  const response = await api.get(`${ATTENDANCE_API}/employee/${employeeId}`);
  return response.data;
};

export const getAttendanceByDateRange = async (startDate, endDate) => {
  const response = await api.get(`${ATTENDANCE_API}/date-range`, {
    params: { startDate, endDate },
  });
  return response.data;
};

export const createAttendance = async (attendanceData) => {
  const response = await api.post(ATTENDANCE_API, attendanceData);
  return response.data;
};

export const updateAttendance = async (id, attendanceData) => {
  const response = await api.put(`${ATTENDANCE_API}/${id}`, attendanceData);
  return response.data;
};

export const deleteAttendance = async (id) => {
  const response = await api.delete(`${ATTENDANCE_API}/${id}`);
  return response.data;
};
